### DISCUSSION TITLE
`man bash` spewing `unsupported control character 0x7` errors on Monterey 12.0.1

### DISCUSSION THREAD

**Discussion by luckman212** [Homebrew/discussions#2506] [Everyday usage]
Since the issue tracker [seems to be down](https://github.com/Homebrew/discussions/discussions/2505), I guess I'll just post this here in case anyone else has seen it.

I noticed that on both of my Macs (one M1 Mini and one Intel MacBook Air) both running 12.0.1, and fully updated brew, I'm getting the following error when running `man bash`:

```
mandoc: /opt/homebrew/share/man/man1/bash.1:34:28: UNSUPP: unsupported control character: 0x7
mandoc: /opt/homebrew/share/man/man1/bash.1:34:35: UNSUPP: unsupported control character: 0x7
mandoc: /opt/homebrew/share/man/man1/bash.1:36:17: UNSUPP: unsupported control character: 0x7
mandoc: /opt/homebrew/share/man/man1/bash.1:36:34: UNSUPP: unsupported control character: 0x7
```

The `/opt/homebrew/share/man/man1/bash.1` file definitely _does_ contain that `^G` character, but it seems to have been there forever, so not sure why it's suddenly causing errors:
![CleanShot 2021-11-22 at 09 14 23@2x](https://user-images.githubusercontent.com/1992842/142876856-3beb5a2d-1ef3-487c-9a24-36f09d7d05ee.png)

Here's a video:

https://user-images.githubusercontent.com/1992842/142876075-ae018e5e-b823-4589-95d4-3d0a4ad78d9e.mp4

I posted a [bugreport](https://savannah.gnu.org/support/?110564) on the GNU Bash homepage, and it was basically pointed back to Apple.

Anyone else seeing this? Is this really a Monterey issue?

```
$ bash --version
GNU bash, version 5.1.12(1)-release (aarch64-apple-darwin21.1.0)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

$ brew config
HOMEBREW_VERSION: 3.3.4-90-g354718c
ORIGIN: https://github.com/Homebrew/brew
HEAD: 354718c40a4e7cc4e5d1d4b03fdce8b992e04e8a
Last commit: 59 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 30877afb911ea723000c2521fd1882cee8906b2d
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_EDITOR: /usr/local/bin/subl -w
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 13.0.0 build 1300
Git: 2.34.0 => /opt/homebrew/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0.1-arm64
CLT: 13.1.0.0.1.1633545042
Xcode: 13.1
Rosetta 2: false

$ brew doctor
Your system is ready to brew.

$ sw_vers
ProductName:	macOS
ProductVersion:	12.0.1
BuildVersion:	21A559
```

**Comment by carlocab** [comment]
I'm seeing this error too. And not just with the `bash` manpage. I see this from `man brew` as well.

> Is this really a Monterey issue?

Probably. Apple seems to have messed up `man` a bit on Monterey. For example, you see this on Monterey ARM:
```
❯ man man
sh: /usr/bin/mandoc: Bad CPU type in executable
```

None of these errors seem to cause any real problems, but it would be nice to not see them.

Could you try opening a ticket at https://feedbackassistant.apple.com?

**Comment by luckman212** [comment]
@carlocab On my system (M1 Mini @ 12.0.1) `man man` does not generate that error. I'll file a ticket anyway

**Reply by carlocab** [reply]
Do you have Rosetta installed? You'll only see it if you don't have Rosetta installed. If you do have Rosetta installed, everything works fine, since
```
❯ file /usr/bin/mandoc
/usr/bin/mandoc: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>
```

**Reply by luckman212** [reply]
Ah yes I do have Rosetta.

```
 $ file /usr/bin/mandoc
/usr/bin/mandoc: Mach-O 64-bit executable x86_64
```

**Reply by carlocab** [reply]
I wouldn't hold my breath for a timely fix: https://twitter.com/zhuowei/status/1429531764503023620

Still, they should hopefully fix this eventually with enough bug reports.

**Reply by luckman212** [reply]
I filed `FB9775632` on the issue

**Reply by oldsharp** [reply]
The brew version gnu-screen man-page produces massive error msg like this, too.

**Comment by luckman212** [comment]
Funny thing is, if I check what command is being executed by `man bash` (using `ps w | grep '[m]an'`) it's this:

```
sh -c '(cd "/opt/homebrew/share/man" && (echo ".ll 10.8i"; echo ".nr LL 10.8i"; /bin/cat '/opt/homebrew/share/man/man1/bash.1') | /usr/bin/tbl | /usr/bin/groff -Wall -mtty-char -Tascii -mandoc -c | (/usr/bin/less -is || true))'
```

And when I run that **exact** command _manually_ it runs just fine, no errors output at all. So, more head scratching. 🤷

**Reply by Alexandreee100** [reply]
Did you solve the problem?

**Reply by luckman212** [reply]
@Alexandreee100 No, and it's not fixed in 12.2 either.

**Reply by Alexandreee100** [reply]
if you find a solution to the problem, can you let me know?

**Comment by scotthesterberg** [comment]
I am seeing this as well, does not seem to fit with the currently allowed sort of issues to report for homebrew:

```
man ggrep
mandoc: /opt/homebrew/share/man/man1/ggrep.1:41:9: UNSUPP: unsupported roff request: do
mandoc: /opt/homebrew/share/man/man1/ggrep.1:42:1: UNSUPP: unsupported roff request: do
mandoc: /opt/homebrew/share/man/man1/ggrep.1:43:1: UNSUPP: unsupported roff request: do
```

**Comment by reegnz** [comment]
I'm experiencing the issue on a 2019 Intel Macbook Pro, so it's not M1 related, Monterey in general seems to be affected (I'm on 12.3.1).

```bash
❯ man bash
mandoc: /usr/local/share/man/man1/bash.1:34:28: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:34:35: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:36:17: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:36:34: UNSUPP: unsupported control character: 0x7

```

It does show the manpage, but I'm seeing this as the first few lines on the terminal before man page is shown.
that's a man page for bash that is installed with homebrew.

**Comment by thomas-merz** [comment]
Same problem here on `GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin21.1.0)` running on a much newer macOS than initially reported:
> System Version:	macOS 12.4 (21F79)
  Kernel Version:	Darwin 21.5.0


When and how will this get fixed? 🤔

**Reply by SMillerDev** [reply]
When enough people report the bug to Apple I guess.

**Comment by shpankau** [comment]
I noticed similar messages this morning after running `man bash` on a 2019 Intel Macbook Pro:
```
$ man bash
mandoc: /usr/local/share/man/man1/bash.1:34:28: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:34:35: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:36:17: UNSUPP: unsupported control character: 0x7
mandoc: /usr/local/share/man/man1/bash.1:36:34: UNSUPP: unsupported control character: 0x7

$ echo ${BASH_VERSION}; uname -a
5.1.16(1)-release
Darwin C02FC5A8MD6M 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64 x86_64 i386 Darwin

$ ls -l $(which man)
-r-xr-xr-x 1 root wheel 239232 Mar 26 03:21 /usr/bin/man

$ realpath $(man -w bash)
/usr/local/Cellar/bash/5.1.16/share/man/man1/bash.1
```
Interestingly, I don't see the messages when running nroff or groff to dump the bash manpage as formatted text, so it's definitely something the macOS man binary doesn't like.

**Comment by carlocab** [comment]
Definitely not a Homebrew issue. It happens on the system-provided `bash` man page too:
```
❯ man bash
mandoc: /usr/share/man/man1/bash.1:35:24: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:35:31: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:37:13: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:37:30: UNSUPP: unsupported control character: 0x7
```

**Reply by luckman212** [reply]
Is there anyone from Apple lurking here? I don't have a developer license so I can't file real bug reports. The ones sent to https://www.apple.com/feedback/macos.html I am pretty sure get routed directly to `/dev/null` (but I filed one anyway)

**Comment by luckman212** [comment]
Well, this is a surprise! Just got this back from Apple to my feedback... looks like it _might_ be fixed in macOS 13 "Ventura". 🎉

![image](https://user-images.githubusercontent.com/1992842/172243030-bdb662c6-8022-4786-bb21-32c9af7fbcee.png)

![image](https://user-images.githubusercontent.com/1992842/172243034-e511848f-0c02-4571-872a-cd889f38f99d.png)

**Comment by ryenus** [comment]
Similarly, when running `/usr/bin/man ggrep`, internally there's below error:

```
$ /usr/bin/zcat -f /opt/homebrew/share/man/man1/ggrep.1 | mandoc -Tlint -Wunsupp
mandoc: <stdin>:41:9: UNSUPP: unsupported roff request: do
mandoc: <stdin>:42:1: UNSUPP: unsupported roff request: do
mandoc: <stdin>:43:1: UNSUPP: unsupported roff request: do
```

And the error:

> This manpage is not compatible with mandoc(1) and might display incorrectly.

**OS**: macOS 13.1

**Related**: [bug#31228: Update GREP manpage](https://www.mail-archive.com/search?l=bug-grep@gnu.org&q=subject:%22bug%2331228%5C%3A+Update+GREP+manpage%22&o=newest&f=1)

**Comment by luckman212** [comment]
As of macOS 13.2.1, the error message has changed

```shell
$ man bash
This manpage is not compatible with mandoc(1) and might display incorrectly.
```

**Comment by 0-issue** [comment]
@luckman212  There is a workaround that I just tried, and it works. Btw, this happens with most man pages with tables too.

Luckily for us, on macOS ``/usr/bin/man`` is not a binary but a shell script. The warning and 1 second delay is coming from that shell script, and can easily be suppressed. Here are the steps:

* Copy ``/usr/bin/man`` to another location on PATH like ``/usr/local/bin`` or ``~/.local/bin/``.
* Open a the copied ``man`` shell script in an editor.
* Search and comment line with ``sleep 1``, and save the file.
* Make sure the file is on PATH before /usr/local/bin
* Open a new terminal session, and try ``man bash`` or whatever previously gave you that error.
* It should work!

```
echo "This manpage is not compatible with mandoc(1) and might display incorrectly." >&2

# Leave sufficient time to note the above warning.
sleep 1 # <- COMMENT this
```

You could comment out ``echo "This manpage is not...`` too though it won't be noticeable, I am keeping that uncommented.

**Reply by thomas-merz** [reply]
This is a not a good workaround… suppressing this message by leaving not sufficient time to display it 🤔 I hoped there's a way to fix it "locally" before Apple will fix the root cause. This would be a nicer workaround!

**Reply by 0-issue** [reply]
It is a good workaround if one knows the man page they are interested in renders well, but they don't want man to keep adding an annoying delay and message. A user can always test if the man page they are viewing has any problem by either scrolling up to look for this line or by invoking man as ``/usr/bin/man <cmdname>``.

**Reply by thomas-merz** [reply]
I understand from your point of view… But why didn't you also remove this line, too? 😉 
> echo "This manpage is not compatible with mandoc(1) and might display incorrectly." >&2

**Reply by 0-issue** [reply]
@thomas-merz You didn't read the last line of my comment. Repeating it here:

> You could comment out ``echo "This manpage is not...`` too though it won't be noticeable, I am keeping that uncommented.

It won't be viewed by the the user when they open a man page greater in height than terminal height as the pager (``less`` in my case) would occupy whole screen upon launch. After quitting man (press ``q`` for less), one can still scroll up to see this line (at least in my terminal setup).

**Reply by 0-issue** [reply]
Btw, here's an issue I filed on asciidoc's generator for roff. It shows a simple example with a manpage table, and what bytes are output by ``man`` vs ``gman`` (gnu man) on macOS: https://github.com/asciidoctor/asciidoctor/issues/4528. Pretty much all the man pages with tables have 1 second delay in opening with message ``This manpage is not compatible with mandoc(1) and might display incorrectly.``.

**Comment by ryenus** [comment]
FYI, if you want to simply remove the unsupported `0x7` characters, try this

```sh
gsed -i '' 's/\x07//g' /opt/homebrew/share/man/man1/bash.1 /opt/homebrew/share/man/man3/expr.3
```

You may consider making it an alias or function for later use.

**Comment by luckman212** [comment]
Been updating this bugreport steadily for 3 years now hoping someone @ Apple will eventually see it. Sadly even as of the upcoming 14.4 release, still not fixed. Would be amazing if others could report as well and reference FB9775632.

**Comment by oldsharp** [comment]
Think I've figured out what's really going on.  (My system is macOS Monterey 12.7.3 so the error msgs are the original mandoc UNSUPP type.)

Root cause
----
Run `man -d /usr/share/man/man1/bash.1` (`man(1)`'s debug mode) and the last 3 lines of the output are (note added by me):
```
man: 
not executing command:
  /usr/bin/mandoc -W unsupp,stop /usr/share/man/man1/bash.1 2>&1 >/dev/null
                                                      ~~~~~~^^^^^^^^^^^^^^^~~~~~~
                                                         This is the root cause
```

Run `/usr/bin/mandoc -W unsupp,stop /usr/share/man/man1/bash.1 2>&1 >/dev/null` and the error msgs are reproduced:
```
mandoc: /usr/share/man/man1/bash.1:35:24: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:35:31: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:37:13: UNSUPP: unsupported control character: 0x7
mandoc: /usr/share/man/man1/bash.1:37:30: UNSUPP: unsupported control character: 0x7
```

Since Apple releases macOS's source code at https://opensource.apple.com/releases/ , the issue can be verified at the code: https://github.com/apple-oss-distributions/man/blob/man-30/man/src/man.c#L774

```c
    char *test_command = my_xsprintf("%s -W unsupp,stop %s 2>&1 >/dev/null",
	"/usr/bin/mandoc", file);

```

It's definitely introduced by Apple since it's inside a `#ifdef __APPLE__` code block.

What's going on
----
macOS Monterey ships with a customized `man(1)` utility (based on the `man(1)` utility of an old FreeBSD release, maybe 7.x or 8.x),  which roughly does these steps during a typical manual pages formatting & displaying process:
1. Run a diagnosis process by doing `/usr/bin/mandoc -W unsupp,stop`.
2. If `mandoc(1)` reports back with an OK, continue formatting with mandoc.
3. Otherwise, formatting with NROFF, which is defined by `man.conf(5)`.

So here at step 1, if `mandoc(1)` encounters an unsupported roff macro it will return to `man(1)` with a non-zero return code, and discard all the error output by FD redirecting & duplicating.  However, the order of FD redirecting & duplicating is wrong. And the message reported by `mandoc(1)` can still be seen on the standard output.

It is a common mistake for shell programming. One could find a detailed explanation at http://mywiki.wooledge.org/BashFAQ/055 .

Some related information taken from the `mandoc(1)` manual page:
```
     -W level
             Specify the minimum message level to be reported on the standard
             error output and to affect the exit status.  The level can be base,
             style, warning, error, or unsupp.  The base level automatically
             derives the operating system from the contents of the Os macro,
             from the -Ios command line option, or from the uname(3) return
             value.  The levels openbsd and netbsd are variants of base that
             bypass autodetection and request validation of base system
             conventions for a particular operating system.  The level all is an
             alias for base.  By default, mandoc is silent.  See EXIT STATUS and
             DIAGNOSTICS for details.

             The special option -W stop tells mandoc to exit after parsing a
             file that causes warnings or errors of at least the requested
             level.  No formatted output will be produced from that file.  If
             both a level and stop are requested, they can be joined with a
             comma, for example -W error,stop.
```
```
EXIT STATUS
     The mandoc utility exits with one of the following values, controlled by
     the message level associated with the -W option:

    ...

     4       At least one unsupported feature was encountered, and -W unsupp or
             a lower level was requested.
    ...
```
```
DIAGNOSTICS
     Messages displayed by mandoc follow this format:

           mandoc: file:line:column: level: message: macro arguments (os)

    ...

     Message levels have the following meanings:

    ...

     unsupp   An input file uses unsupported low-level roff(7) features.  The
              output may be incomplete and/or misformatted, so using GNU troff
              instead of mandoc to process the file may be preferable.
```

This also explains why the command for formatting & displaying a bash manual page like this (see by `ps(1)` during a manual page display) does not show any error message:
```
/usr/bin/tbl /usr/share/man/man1/bash.1 | /usr/bin/groff -Wall -mtty-char -Tascii -mandoc -c | less -is
```
because this is the step 3.  One can further verify it by doing a `man man` and `ps -ef | grep mandoc`:
```
sh -c (cd /usr/share/man && /usr/bin/mandoc -c -O width=80 '/usr/share/man/man1/man.1') | (/usr/bin/less -is || true)
```
This time the step 2 is chosen.

How to fix
----

### for macOS Monterey
While workaround may vary, the real fix should take place at the `man(1)` utility shipped with macOS.  Either `>/dev/null 2>&1` or simply `2>/dev/null` should do the trick.

Apply the patch below to the source tree https://github.com/apple-oss-distributions/man/tree/man-30 should fix the problem:

```patch
From 7f9e33f21cdd7fc930c064e709ee297693ee3b9e Mon Sep 17 00:00:00 2001
From: Rui Chen <rui@hexpunks.org>
Date: Mon, 12 Feb 2024 22:58:52 +0800
Subject: [PATCH 1/1] Fix mandoc unsupp warning msg

---
 man/src/man.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/man/src/man.c b/man/src/man.c
index f3a7246..d934ca1 100644
--- a/man/src/man.c
+++ b/man/src/man.c
@@ -771,7 +771,7 @@ make_mandoc_command(const char * path, const char *file) {
 	return NULL;
     }
 
-    char *test_command = my_xsprintf("%s -W unsupp,stop %s 2>&1 >/dev/null",
+    char *test_command = my_xsprintf("%s -W unsupp,stop %s >/dev/null 2>&1",
 	"/usr/bin/mandoc", file);
     int test_status = do_system_command(test_command, 1);
     if (test_status == 0) {
-- 
2.37.1 (Apple Git-137.1)


```

Unfortunately,  compile process isn't as straight; I am still struggling at it.

### for macOS 13 and newer
Starting from macOS 13, Apple ships its system with a `man(1)` utility from a newer FreeBSD release (maybe 13.2). It's a  shell script. So the fix no longer apply.

**Comment by luckman212** [comment]
Apple recently updated my feedback FB9775632 on this to read _"Potential fix identified - For a future OS update"_

<img width=463 src=https://github.com/Homebrew/discussions/assets/1992842/b80b20d3-0db5-4424-9b85-c61c18f57776>

So, a fix may be forthcoming.

**Comment by sebma** [comment]
To fix the bash manual error, I typed this : 
```shell
gsed -i 's/\x07//g' $(man -w bash)
```

**Comment by thomas-merz** [comment]
At least with a more current MacOS (14.7.2) this is "fixed" with this message/info:

> This manpage is not compatible with mandoc(1) and might display incorrectly.
